api/envoy/extensions/filters/http/local_ratelimit/v3/local_rate_limit.proto

```diff
@@ -67,4 +79,19 @@ message LocalRateLimit {
   // have been rate limited.
   repeated config.core.v3.HeaderValueOption response_headers_to_add = 6
       [(validate.rules).repeated = {max_items: 10}];
+
+  // The rate limit descriptor list to use in the local rate limit to override
+  // on.
+  repeated common.ratelimit.v3.RateLimitDescriptor descriptors = 8;
```

Comment:
From a user perspective I'm kind of confused about how this is going to be used. Can you add a fully worked example either here or in the config/arch docs somewhere that makes it easier to understand how/why the user would configure this? It seems like we are just doing string matching here but I'm not exactly sure what we are matching against. Thank you!

Reply:
Thanks @mattklein123  and @rgs1  for the review! 
This is the example, I want to get feedback first before I do refactoring to make code common with Http Global filter. 
Taking the yaml from the test and editing it a little bit. 

stat_prefix: test
token_bucket:
  max_tokens: {}
  tokens_per_fill: 1
  fill_interval: 1000s
filter_enabled:
  runtime_key: test_enabled
  default_value:
    numerator: 100
    denominator: HUNDRED
filter_enforced:
  runtime_key: test_enforced
  default_value:
    numerator: 100
    denominator: HUNDRED
response_headers_to_add:
  - append: false
    header:
      key: x-test-rate-limit
      value: 'true'
descriptors:
- entries:
   - key: client_id
     value: foo
   - key: path
     value: /foo/bar
   limit: 
     requests_per_unit: 10
     unit: MINUTE
- entries:
   - key: client_id
     value: foo
   limit: 
     requests_per_unit: 100
     unit: MINUTE

What this says is that for any route, we allow 1000request/min but if "client_id" is "foo" and path is "foo/bar" we only allow 10 requests per min and if just "client_id" is "foo" then we allow 100 requests per minute. As said in the description above, I am thinking to map request_per_unit to tokens_per_fill and unit to fill_interval in token bucket algo implementation. Does that sound ok?

Reply:
Can you format the YAML using code blocks and also provide the parallel route config that matches it?

Reply:
```yaml
stat_prefix:` test
token_bucket:
  max_tokens: {}
  tokens_per_fill: 1
  fill_interval: 1000s
filter_enabled:
  runtime_key: test_enabled
  default_value:
    numerator: 100
    denominator: HUNDRED
filter_enforced:
  runtime_key: test_enforced
  default_value:
    numerator: 100
    denominator: HUNDRED
response_headers_to_add:
  - append: false
    header:
      key: x-test-rate-limit
      value: 'true'
descriptors:
- entries:
   - key: client_id
     value: foo
   - key: path
     value: /foo/bar
   limit: 
     requests_per_unit: 10
     unit: MINUTE
- entries:
   - key: client_id
     value: foo
   limit: 
     requests_per_unit: 100
     unit: MINUTE
```

Reply:
@mattklein123 : friendly ping... ptal :)

Reply:
Please show me the companion rate limit config in the route. What you have above isn't actually going to do anything AFAICT. Thank you.

Reply:
ahaa thanks.. added.. please see if this looks accurate..

```yaml
route_config:
  name: local_route
  virtual_hosts:
  - name: local_service
    domains: ["*"]
    routes:
    - match: { prefix: "/foo" }
      route: { cluster: service_protected_by_rate_limit }
      typed_per_filter_config:
        envoy.filters.http.local_ratelimit:
          "@type": type.googleapis.com/envoy.extensions.filters.http.local_ratelimit.v3.LocalRateLimit
          stat_prefix: test
          token_bucket:
            max_tokens: 1000
            tokens_per_fill: 1000
            fill_interval: 60s
          filter_enabled:
            runtime_key: test_enabled
            default_value:
              numerator: 100
              denominator: HUNDRED
          filter_enforced:
            runtime_key: test_enforced
            default_value:
              numerator: 100
              denominator: HUNDRED
          response_headers_to_add:
            - append: false
              header:
                key: x-test-rate-limit
                value: 'true'
          descriptors:
          - entries:
            - key: client_id
              value: foo
            - key: path
              value: /foo/bar
            limit: 
              requests_per_unit: 10
              unit: MINUTE
          - entries:
            - key: client_id
              value: foo
            limit: 
              requests_per_unit: 100
              unit: MINUTE
    - match: { prefix: "/" }
      route: { cluster: default_service }
    rate_limits:
    - actions: # any actions in here
      - request_headers:
          header_name: ":path"
          descriptor_key: "PATH"
      - generic_key:
          descriptor_value: "foo"
          descriptor_key: "client_id"
    - actions: 
      - generic_key:
          descriptor_value: "foo"
          descriptor_key: "client_id"
```

Reply:
Yeah at a high level this looks right to me, and I agree that we could use the override limits from the descriptors potentially. Do you want to flesh out the change more and we can circle back?

Reply:
Cool.. thanks for getting back!
Will update the PR with the changes and let you guys know.

Reply:
@mattklein123 : This is ready for review again